File: server/lib/silenceGenerator.js
import logger from '../utils/logger.js';

class SilenceGenerator {
    /**
     * Generate silence MP3 frames
     * @param {number} durationMs - Duration in milliseconds
     * @param {number} bitrate - Bitrate in kbps (default 128)
     * @param {number} sampleRate - Sample rate in Hz (default 44100)
     * @returns {Buffer} Buffer containing silence MP3 frames
     */
    static generateSilence(durationMs = 100, bitrate = 128, sampleRate = 44100) {
        try {
            // MP3 frame size calculation
            const samplesPerFrame = 1152; // Standard for MPEG-1 Layer III
            const frameDuration = (samplesPerFrame / sampleRate) * 1000; // ms per frame
            const frameCount = Math.ceil(durationMs / frameDuration);
            
            // Create a buffer for silence frames
            const frames = [];
            
            // MP3 frame header for silence (MPEG-1 Layer III)
            const frameHeader = Buffer.from([
                0xFF, 0xFB, 
                0x90, 0x00
            ]);
            
            // Calculate frame size
            const frameSize = Math.floor((144 * bitrate * 1000) / sampleRate);
            
            // Generate silence frames
            for (let i = 0; i < frameCount; i++) {
                const frame = Buffer.alloc(frameSize);
                frameHeader.copy(frame, 0);
                frames.push(frame);
            }
            
            const silenceBuffer = Buffer.concat(frames);
            logger.debug(`Generated ${durationMs}ms of silence (${silenceBuffer.length} bytes)`);
            
            return silenceBuffer;
        } catch (error) {
            logger.error('Error generating silence:', error);
            return Buffer.from([0xFF, 0xFB, 0x90, 0x00]);
        }
    }
    
    /**
     * Generate a fade-in or fade-out buffer
     * @param {Buffer} audioBuffer - The audio buffer to fade
     * @param {string} type - 'in' or 'out'
     * @param {number} durationMs - Duration of fade in milliseconds
     * @returns {Buffer} Buffer with fade applied
     */
    static applyFade(audioBuffer, type = 'in', durationMs = 50) {
        try {
            return audioBuffer;
        } catch (error) {
            logger.error('Error applying fade:', error);
            return audioBuffer;
        }
    }
}

export default SilenceGenerator;

File: app/streaming/silence_generator.py
from app.core import logger

class SilenceGenerator:
    def __init__(self):
        self.bitrate = 128
        self.sample_rate = 44100
    
    def generate_silence(self):
        """Generate infinite silence stream as an iterator"""
        try:
            frame_header = bytes([0xFF, 0xFB, 0x90, 0x00])
            frame_size = int((144 * self.bitrate * 1000) / self.sample_rate)
            
            silence_frame = bytearray(frame_size)
            silence_frame[0:4] = frame_header
            silence_bytes = bytes(silence_frame)
            
            logger.info(f"Starting silence generation (frame size: {frame_size} bytes)")
            
            while True:
                yield silence_bytes
        except Exception as error:
            logger.error(f'Error generating silence: {str(error)}')
            yield bytes([0xFF, 0xFB, 0x90, 0x00])
    
    @staticmethod
    def apply_fade(audio_buffer: bytes, fade_type: str = 'in', duration_ms: int = 50) -> bytes:
        try:
            return audio_buffer
        except Exception as error:
            logger.error(f'Error applying fade: {str(error)}')
            return audio_buffer

silence_generator = SilenceGenerator()
